File: android/truepad-storage/src/main/kotlin/dev/systemslibrarian/truepad/storage/Witness.kt
package dev.systemslibrarian.truepad.storage

import dev.systemslibrarian.truepad.core.Direction
import dev.systemslibrarian.truepad.core.JsonNumber
import dev.systemslibrarian.truepad.core.JsonObject
import dev.systemslibrarian.truepad.core.JsonString
import dev.systemslibrarian.truepad.core.JsonValue
import dev.systemslibrarian.truepad.core.parseJson

/*
 * The crash-safe Android rollback witness — the Kotlin twin of
 * src/browser/engine/witness.ts (FORMAT-V2.md §15).
 *
 * The witness remembers how far a store has advanced, so a store rolled back by
 * a backup/device-transfer restore refuses to move (`witness-regressed`) instead
 * of reusing retired positions or refilling a contested record's attempt budget.
 * It records EXACTLY the three frozen monotone counters and nothing else (§15.1):
 *
 *   { encryptionNextOffset, authenticationNextSequence, attemptsReserved }
 *
 * This is an ANDROID-PRODUCT layer, NOT a class of the frozen store. head.json
 * always carries the CLI's rollback:{witnessClass:"none"} (Store.kt); whether a
 * pair also carries an Android-local witness is recorded in the Android-only
 * pair.json (`witness`). The two kinds mirror the Browser Edition's exactly:
 *
 *   android-none          — no witness. A no-op. Restoring app-private storage
 *                           regresses the store and resets the per-record
 *                           attempt budget; docs/ANDROID-SECURITY.md says so.
 *                           A bare FORMAT-V2 store this app never provisioned
 *                           (e.g. a CLI store copied in) is android-none.
 *   android-local-witness — an APPEND-ONLY journal at `witness/<pairId>.log`,
 *                           written through the Engine's SEPARATE witness Fs.
 *
 * WHERE THE JOURNAL LIVES IS THE WHOLE POINT. A witness only detects a rollback
 * if it sits in a different failure domain from the thing being rolled back. The
 * Android app therefore puts the store under `Context.getFilesDir()` and this
 * journal under `Context.getNoBackupFilesDir()`: Android Auto Backup and
 * device-to-device transfer carry the former and NOT the latter, so a restored
 * pair store meets a witness that still remembers the true high-water and the
 * true attempt budget, and refuses `witness-regressed` before anything is
 * consumed.
 *
 * Named honestly: android-LOCAL. It is one directory on one device, not an
 * independent host, and it claims nothing more. Uninstall and "Clear storage"
 * take both trees — and that is LOSS, not reuse, which is the trade this product
 * always makes. An attacker who can already rewrite this app's private storage
 * is outside what any local witness can defend against.
 *
 * CRASH SAFETY. The journal is APPEND-ONLY and is NEVER truncated, so no write
 * interruption can shrink it, and records are LEADING-newline framed
 * (`\n<json>`, encodeRecord). appendFile gives no record boundary, so a crash
 * mid-append can leave a newline-free partial at EOF — but leading framing
 * bounds every record by its own `\n` and the next record's `\n`, so a torn
 * partial is always an isolated line, never fused with the record before or
 * after it. The read DROPS any line that does not parse and folds the SURVIVING
 * records into the per-direction elementwise max. So only a TORN advance loses
 * its own value; every advance whose append COMPLETED is preserved. Crucially, a
 * torn advance's operation ERRORED and WITHHELD its output (burn emits the
 * envelope / open releases the plaintext only AFTER a successful advance), so
 * the witness never under-reports below a state whose output was RELEASED, and
 * the very next clean advance re-records the current high-water (self-heal).
 *
 * Because a provisioned journal is never emptied by an advance, an established
 * android-local witness NEVER reads as "fresh": a provisioned pair whose journal
 * is missing, empty, all-corrupt, or missing a direction fails CLOSED as
 * `witness-inconsistent`. Bootstrap — the explicit provisioning event — is the
 * only writer that creates the first records, at gen or a successful import,
 * never inferred from an empty file.
 *
 * (Skipping — not failing closed on — a malformed line is safe: the witness
 * cannot defend against an attacker who can already rewrite this app's private
 * storage. Its jobs are crash-safety and detecting a rollback of the PAIR store.)
 */

/** The Android-product witness kind, as recorded in pair.json. */
enum class WitnessKind(val wire: String) {
    NONE("android-none"),
    LOCAL("android-local-witness");

    companion object {
        fun fromWire(s: String?): WitnessKind? = when (s) {
            "android-none" -> NONE
            "android-local-witness" -> LOCAL
            else -> null
        }
    }
}

/** The three frozen monotone counters, and nothing else (§15.1, ledger N17). */
data class WitnessCounters(
    val encryptionNextOffset: Long,
    val authenticationNextSequence: Long,
    val attemptsReserved: Long,
) {
    companion object {
        val ZERO = WitnessCounters(0, 0, 0)
    }
}

/** The non-secret comparison of a store against its witness, for status (§15.3). */
enum class WitnessState(val wire: String) {
    NA("n/a"), ALIGNED("aligned"), AHEAD("ahead"), REGRESSED("regressed"), INCONSISTENT("inconsistent")
}

sealed class WitnessPreflight {
    data class Ok(val state: WitnessState) : WitnessPreflight()
    /** reason ∈ {"witness-regressed","witness-inconsistent"}. */
    data class Refusal(val reason: String, val message: String) : WitnessPreflight()
}

/** The store's effective high-waters, for the preflight/report comparison. */
data class StoreHighWaters(val nextOffset: Long, val nextSequence: Long, val attemptsReserved: Long)

fun witnessLogPath(pairId: String): String = "witness/$pairId.log"

/* ---- the append-only journal ---------------------------------------------- */

private const val MAX_SAFE = 9_007_199_254_740_991L

private fun safeCount(v: JsonValue?): Long? {
    if (v !is JsonNumber) return null
    val n = v.raw.toLongOrNull() ?: return null
    return if (n in 0..MAX_SAFE) n else null
}

/**
 * Each record is framed with a LEADING newline (`\n<json>`), NOT a trailing one.
 * This is what makes a torn append harmless to its neighbours: appendFile writes
 * at EOF with no boundary, so a crash mid-append can leave a newline-free
 * partial. With leading framing, every record — the torn one included — is
 * bounded on the LEFT by its own `\n` and on the RIGHT by the NEXT record's
 * `\n`, so a torn partial is always an isolated line that the reader drops, and
 * it can never fuse into and destroy the record before OR after it. (Trailing
 * framing would let a torn partial swallow the following clean record.)
 *
 * The four short keys and their order match the Browser Edition byte for byte.
 */
internal fun encodeWitnessRecord(direction: Direction, c: WitnessCounters): ByteArray {
    val sb = StringBuilder("\n{\"d\":")
    jsonString(sb, direction.wire)
    sb.append(",\"eno\":").append(c.encryptionNextOffset)
        .append(",\"ans\":").append(c.authenticationNextSequence)
        .append(",\"ar\":").append(c.attemptsReserved)
        .append('}')
    return sb.toString().toByteArray(Charsets.UTF_8)
}

private class WitnessRecord(val direction: Direction, val counters: WitnessCounters)

private fun parseWitnessRecord(raw: JsonValue): WitnessRecord? {
    if (raw !is JsonObject) return null
    if (raw.members.size != 4) return null
    val direction = Direction.fromWire((raw.members["d"] as? JsonString)?.value ?: "") ?: return null
    val eno = safeCount(raw.members["eno"]) ?: return null
    val ans = safeCount(raw.members["ans"]) ?: return null
    val ar = safeCount(raw.members["ar"]) ?: return null
    return WitnessRecord(direction, WitnessCounters(eno, ans, ar))
}

/**
 * Read the append-only journal and fold the SURVIVING records into the
 * per-direction elementwise maximum. This is what makes the journal crash-safe
 * WITHOUT relying on any atomic replace. `null` means absent / empty / no
 * surviving record — no provisioned high-water.
 */
private fun readEffective(fs: Fs, pairId: String): Map<Direction, WitnessCounters>? {
    val bytes = fs.readFile(witnessLogPath(pairId)) ?: return null
    val byDirection = LinkedHashMap<Direction, WitnessCounters>()
    for (line in String(bytes, Charsets.UTF_8).split("\n")) {
        if (line.isEmpty()) continue
        val parsed = try { parseJson(line) } catch (_: Exception) { continue } // torn/corrupt — drop it
        val record = parseWitnessRecord(parsed) ?: continue
        val prev = byDirection[record.direction]
        byDirection[record.direction] = if (prev == null) {
            record.counters
        } else {
            WitnessCounters(
                maxOf(prev.encryptionNextOffset, record.counters.encryptionNextOffset),
                maxOf(prev.authenticationNextSequence, record.counters.authenticationNextSequence),
                maxOf(prev.attemptsReserved, record.counters.attemptsReserved),
            )
        }
    }
    return if (byDirection.isEmpty()) null else byDirection
}

private fun belowWitness(s: StoreHighWaters, w: WitnessCounters): Boolean =
    s.nextOffset < w.encryptionNextOffset ||
        s.nextSequence < w.authenticationNextSequence ||
        s.attemptsReserved < w.attemptsReserved

private fun alignedWith(s: StoreHighWaters, w: WitnessCounters): Boolean =
    s.nextOffset == w.encryptionNextOffset &&
        s.nextSequence == w.authenticationNextSequence &&
        s.attemptsReserved == w.attemptsReserved

/* ---- the witness contract -------------------------------------------------- */

interface Witness {
    val kind: WitnessKind

    /**
     * Provision the witness — the explicit event at gen or a successful import.
     * `initial` seeds each direction (gen: zeros; import: the imported store's
     * high-waters, so a mid-life import is not spuriously refused
     * witness-regressed). [WitnessKind.NONE] is a no-op.
     */
    fun bootstrap(pairId: String, initial: Map<Direction, WitnessCounters>? = null)

    /**
     * §15.3 PREFLIGHT: a free state gate before anything is consumed. A store
     * below its witness refuses `witness-regressed`; a missing/empty/torn/
     * absent-direction PROVISIONED witness refuses `witness-inconsistent`
     * (fail closed — an established witness never reads as fresh).
     */
    fun preflight(pairId: String, direction: Direction, store: StoreHighWaters): WitnessPreflight

    /**
     * §15.3 ADVANCE: after the durable §12 commit, before the emit. Appends one
     * record. Throws on an I/O failure — the caller has already committed, so
     * the output is withheld (the LOSS row: material, never reuse).
     */
    fun advance(pairId: String, direction: Direction, counters: WitnessCounters)

    /** §15.3 status: read-only comparison; refuses nothing. */
    fun report(pairId: String, direction: Direction, store: StoreHighWaters): WitnessState
}

/** android-none: no witness. Every touchpoint is a no-op; status reports "n/a". */
private object NoneWitness : Witness {
    override val kind = WitnessKind.NONE
    override fun bootstrap(pairId: String, initial: Map<Direction, WitnessCounters>?) { /* nothing to provision */ }
    override fun preflight(pairId: String, direction: Direction, store: StoreHighWaters) =
        WitnessPreflight.Ok(WitnessState.NA)
    override fun advance(pairId: String, direction: Direction, counters: WitnessCounters) { /* nothing to advance */ }
    override fun report(pairId: String, direction: Direction, store: StoreHighWaters) = WitnessState.NA
}

private fun inconsistent(message: String) =
    WitnessPreflight.Refusal("witness-inconsistent", "$message Nothing was burned.")

/** android-local-witness: the append-only journal at `witness/<pairId>.log`. */
private class LocalWitness(private val fs: Fs) : Witness {
    override val kind = WitnessKind.LOCAL

    override fun bootstrap(pairId: String, initial: Map<Direction, WitnessCounters>?) {
        // Provision both directions; protection begins here. Append-only: these
        // two records are the journal's first durable content.
        for (d in listOf(Direction.A_TO_B, Direction.B_TO_A)) {
            val seed = initial?.get(d) ?: WitnessCounters.ZERO
            fs.appendFile(witnessLogPath(pairId), encodeWitnessRecord(d, seed))
        }
    }

    override fun preflight(pairId: String, direction: Direction, store: StoreHighWaters): WitnessPreflight {
        val eff = readEffective(fs, pairId)
            ?: return inconsistent(
                "this pair is provisioned with an Android-local rollback witness, but its journal " +
                    "${witnessLogPath(pairId)} is missing, empty, or holds no surviving record. An established " +
                    "witness is never treated as fresh: a vanished witness is a possible rollback, so this fails closed.",
            )
        val w = eff[direction]
            ?: return inconsistent(
                "the Android-local rollback witness ${witnessLogPath(pairId)} carries no record for " +
                    "${direction.wire}: its provisioning record for this direction is gone, so it fails closed " +
                    "rather than assume a fresh store.",
            )
        if (belowWitness(store, w)) {
            return WitnessPreflight.Refusal(
                "witness-regressed",
                "this store is behind its rollback witness: the witness records encryptionNextOffset " +
                    "${w.encryptionNextOffset}, authenticationNextSequence ${w.authenticationNextSequence}, and " +
                    "attemptsReserved ${w.attemptsReserved}, but this store is at nextOffset ${store.nextOffset}, " +
                    "nextSequence ${store.nextSequence}, and attemptsReserved ${store.attemptsReserved}. A store " +
                    "below its witness is the restored-store signature (§9.4): the pair store was rolled back while " +
                    "the witness, in a separate directory the pair's own files do not include, remembers the true " +
                    "high-water and attempt budget. Refusing before anything is consumed. Nothing was burned.",
            )
        }
        return WitnessPreflight.Ok(if (alignedWith(store, w)) WitnessState.ALIGNED else WitnessState.AHEAD)
    }

    override fun advance(pairId: String, direction: Direction, counters: WitnessCounters) {
        fs.appendFile(witnessLogPath(pairId), encodeWitnessRecord(direction, counters))
    }

    override fun report(pairId: String, direction: Direction, store: StoreHighWaters): WitnessState {
        val eff = readEffective(fs, pairId) ?: return WitnessState.INCONSISTENT
        val w = eff[direction] ?: return WitnessState.INCONSISTENT
        if (belowWitness(store, w)) return WitnessState.REGRESSED
        return if (alignedWith(store, w)) WitnessState.ALIGNED else WitnessState.AHEAD
    }
}

/** Build the witness for a pair's Android-product witness kind (from pair.json). */
fun witnessFor(fs: Fs, kind: WitnessKind): Witness =
    if (kind == WitnessKind.LOCAL) LocalWitness(fs) else NoneWitness

File: ios/TruePadKit/Sources/TruePadStorage/Witness.swift
/* ============================================================================
 * The rollback witness — an append-only journal of three monotone counters.
 * ----------------------------------------------------------------------------
 * Swift twin of android/truepad-storage Witness.kt and src/browser/engine's
 * witness. FORMAT-V2 §15.
 *
 * A restore is the classic pad-reuse vector: put yesterday's store back, and
 * every byte spent since becomes spendable again. Worse, restoring only the
 * header refills a contested record's verification-attempt budget and defeats the
 * finite forgery bound of §5. The witness records exactly three frozen monotone
 * counters per direction — encryptionNextOffset, authenticationNextSequence,
 * attemptsReserved — and nothing else. Never pad contents, keys, masks,
 * plaintext, or ciphertext.
 *
 * A WITNESS ONLY DETECTS A ROLLBACK IF IT IS IN A DIFFERENT FAILURE DOMAIN FROM
 * THE THING BEING ROLLED BACK. That is why the Engine takes a SEPARATE `Fs` for
 * it. Android binds the store to getFilesDir() and the witness to
 * getNoBackupFilesDir(), because Android's backup carries the former and not the
 * latter.
 *
 * iOS HAS NO SUCH DIRECTORY. There is no documented "the backup system skips
 * this folder" root. So on iOS the different domain has to come from somewhere
 * else, and the witness is deliberately written against the small `Fs` surface it
 * actually uses — readFile, appendFile, exists, remove — so that a
 * Keychain-backed conformer can supply that domain without the engine changing.
 * Until the app binds one, `witnessFs` defaults to the store's own `Fs`, which is
 * the honest-but-weak configuration in which a restore carries the witness too
 * and the rollback is invisible. That weak case is TESTED rather than hidden.
 *
 * See docs/IOS-SECURITY.md §5 for what is relied upon and what is not.
 * ========================================================================= */

import Foundation
import TruePadCore

/// The witness kind, as recorded in the platform-local pair metadata — never in
/// the frozen head, which always carries `rollback: { witnessClass: "none" }`.
public enum WitnessKind: String, Sendable, Equatable {
    case none = "ios-none"
    case local = "ios-local-witness"

    public static func fromWire(_ s: String?) -> WitnessKind? {
        guard let s else { return nil }
        return WitnessKind(rawValue: s)
    }
}

/// The three frozen monotone counters, and nothing else (§15.1).
public struct WitnessCounters: Sendable, Equatable {
    public let encryptionNextOffset: Int
    public let authenticationNextSequence: Int
    /// The store's count of `attempt` journal lines. This is the quantity that
    /// stops a restore refilling a contested record's verification budget, which
    /// is why it is witnessed alongside the two obvious cursors.
    public let attemptsReserved: Int

    public init(encryptionNextOffset: Int, authenticationNextSequence: Int, attemptsReserved: Int) {
        self.encryptionNextOffset = encryptionNextOffset
        self.authenticationNextSequence = authenticationNextSequence
        self.attemptsReserved = attemptsReserved
    }

    public static let zero = WitnessCounters(encryptionNextOffset: 0,
                                             authenticationNextSequence: 0,
                                             attemptsReserved: 0)

    func elementwiseMax(_ other: WitnessCounters) -> WitnessCounters {
        WitnessCounters(
            encryptionNextOffset: max(encryptionNextOffset, other.encryptionNextOffset),
            authenticationNextSequence: max(authenticationNextSequence, other.authenticationNextSequence),
            attemptsReserved: max(attemptsReserved, other.attemptsReserved))
    }
}

/// The non-secret comparison of a store against its witness, for status (§15.3).
public enum WitnessState: String, Sendable, Equatable {
    case na = "n/a"
    case aligned = "aligned"
    case ahead = "ahead"
    case regressed = "regressed"
    case inconsistent = "inconsistent"
}

public enum WitnessPreflight: Sendable {
    case ok(WitnessState)
    case refused(reason: String, message: String)
}

/// The store's effective high-waters, for the comparison.
public struct StoreHighWaters: Sendable, Equatable {
    public let nextOffset: Int
    public let nextSequence: Int
    public let attemptsReserved: Int

    public init(nextOffset: Int, nextSequence: Int, attemptsReserved: Int) {
        self.nextOffset = nextOffset
        self.nextSequence = nextSequence
        self.attemptsReserved = attemptsReserved
    }
}

public func witnessLogPath(_ pairId: String) -> String { "witness/\(pairId).log" }

/// Serialize one record with a LEADING newline, four short keys in fixed order.
///
/// THE LEADING NEWLINE IS LOAD-BEARING. `appendFile` writes at EOF with no record
/// boundary, so a crash mid-append leaves a newline-free partial. With leading
/// framing every record — the torn one included — is bounded on the LEFT by its
/// own `\n` and on the RIGHT by the NEXT record's `\n`, so a torn partial is
/// always an isolated line the reader drops, and can never fuse into and destroy
/// the record before or after it. TRAILING framing would let a torn partial
/// swallow the following clean record, which is exactly the crash case this
/// design exists for.
func encodeWitnessRecord(_ direction: PadDirection, _ c: WitnessCounters) -> [UInt8] {
    var s = "\n{\"d\":"
    appendJsonString(&s, direction.rawValue)
    s.append(",\"eno\":\(c.encryptionNextOffset)")
    s.append(",\"ans\":\(c.authenticationNextSequence)")
    s.append(",\"ar\":\(c.attemptsReserved)}")
    return Array(s.utf8)
}

public protocol Witness {
    var kind: WitnessKind { get }

    /// The explicit provisioning event, at gen or a successful import.
    func bootstrap(pairId: String, initial: [PadDirection: WitnessCounters]?) throws

    /// §15.3 PREFLIGHT — a free state gate before anything is consumed.
    func preflight(pairId: String, direction: PadDirection, store: StoreHighWaters) -> WitnessPreflight

    /// §15.3 ADVANCE — after the durable §12 commit, before the emit.
    func advance(pairId: String, direction: PadDirection, counters: WitnessCounters) throws

    /// §15.3 status: read-only comparison. Refuses nothing, ever.
    func report(pairId: String, direction: PadDirection, store: StoreHighWaters) -> WitnessState
}

/// No witness, no claim. A bare FORMAT-V2 store this app never provisioned — a
/// CLI store copied in — is `none`. Restoring it regresses it and resets the
/// attempt budget, and FORMAT-V2 §9.4 stands as written.
public struct NoneWitness: Witness {
    public let kind: WitnessKind = .none
    public init() {}
    public func bootstrap(pairId: String, initial: [PadDirection: WitnessCounters]?) throws {}
    public func preflight(pairId: String, direction: PadDirection,
                          store: StoreHighWaters) -> WitnessPreflight { .ok(.na) }
    public func advance(pairId: String, direction: PadDirection, counters: WitnessCounters) throws {}
    public func report(pairId: String, direction: PadDirection,
                       store: StoreHighWaters) -> WitnessState { .na }
}

public struct LocalWitness: Witness {
    public let kind: WitnessKind = .local
    let fs: Fs

    public init(fs: Fs) { self.fs = fs }

    /// The per-direction elementwise maximum over the SURVIVING records, or nil
    /// when the journal is absent or holds no surviving record.
    ///
    /// A lower value in any record — including a junk or replayed line — can
    /// never pull the fold down, because the fold is a maximum.
    func readEffective(_ pairId: String) -> [PadDirection: WitnessCounters]? {
        guard let bytes = ((try? fs.readFile(witnessLogPath(pairId))) ?? nil) else { return nil }
        let text = String(decoding: bytes, as: UTF8.self)
        var out: [PadDirection: WitnessCounters] = [:]

        for line in text.components(separatedBy: "\n") {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.isEmpty { continue }
            // Malformed lines FAIL OPEN — they are skipped. That is safe because
            // the witness cannot defend against an attacker who can already
            // rewrite this app's private storage; its jobs are crash-safety and
            // detecting a rollback of the PAIR store. A missing/empty/all-corrupt
            // journal on a PROVISIONED pair is the opposite case and fails closed.
            guard let parsed = try? parseStrictJson(trimmed),
                  case .object(let members) = parsed,
                  members.count == 4,                       // EXACTLY four keys
                  let map = parsed.memberMap,
                  case .string(let dirText)? = map["d"],
                  let direction = PadDirection.fromWire(dirText),
                  let eno = safeCount(map["eno"]),
                  let ans = safeCount(map["ans"]),
                  let ar = safeCount(map["ar"]) else { continue }

            let record = WitnessCounters(encryptionNextOffset: eno,
                                         authenticationNextSequence: ans,
                                         attemptsReserved: ar)
            out[direction] = out[direction].map { $0.elementwiseMax(record) } ?? record
        }
        return out.isEmpty ? nil : out
    }

    /// A non-negative safe integer in its RAW spelling. `1.0` and `1e3` are
    /// rejected — the record is dropped — even though JavaScript would read them
    /// as integer-valued. The parser keeps the raw text precisely so nothing is
    /// folded away.
    func safeCount(_ v: JsonValue?) -> Int? {
        guard case .number(let raw)? = v, let n = Int(raw), n >= 0, n <= maxSafeInteger else {
            return nil
        }
        return n
    }

    static func belowWitness(_ s: StoreHighWaters, _ w: WitnessCounters) -> Bool {
        s.nextOffset < w.encryptionNextOffset
            || s.nextSequence < w.authenticationNextSequence
            || s.attemptsReserved < w.attemptsReserved
    }

    static func alignedWith(_ s: StoreHighWaters, _ w: WitnessCounters) -> Bool {
        s.nextOffset == w.encryptionNextOffset
            && s.nextSequence == w.authenticationNextSequence
            && s.attemptsReserved == w.attemptsReserved
    }

    /// Both inconsistency messages get " Nothing was burned." appended — a space
    /// then the sentence. The regression message carries it inside its own
    /// literal instead, so a port must not append it uniformly or it will be
    /// doubled.
    static func inconsistent(_ message: String) -> WitnessPreflight {
        .refused(reason: "witness-inconsistent", message: "\(message) Nothing was burned.")
    }

    public func bootstrap(pairId: String, initial: [PadDirection: WitnessCounters]? = nil) throws {
        // Exactly two records, A->B then B->A. Bootstrap is the ONLY writer that
        // creates the first records: freshness is never inferred from an empty or
        // absent file.
        for direction in [PadDirection.aToB, PadDirection.bToA] {
            let counters = initial?[direction] ?? .zero
            try fs.appendFile(witnessLogPath(pairId), encodeWitnessRecord(direction, counters))
        }
    }

    public func preflight(pairId: String, direction: PadDirection,
                          store: StoreHighWaters) -> WitnessPreflight {
        guard let effective = readEffective(pairId) else {
            // Absent, empty, or no surviving record — one case, one message. An
            // established witness NEVER reads as fresh, because a vanished
            // witness is indistinguishable from a rollback that took it.
            return Self.inconsistent(
                "\(witnessLogPath(pairId)) is missing or holds no readable record, but this pair is "
                + "provisioned with a local rollback witness. A provisioned witness is never emptied "
                + "by normal use, so this is either a restore that dropped it or damage; either way "
                + "the pair cannot be shown to have moved forward.")
        }
        guard let w = effective[direction] else {
            return Self.inconsistent(
                "\(witnessLogPath(pairId)) holds no record for \(direction.rawValue), but this pair "
                + "is provisioned with a local rollback witness. This direction cannot be shown to "
                + "have moved forward.")
        }
        if Self.belowWitness(store, w) {
            return .refused(reason: "witness-regressed", message:
                "This store is BEHIND its rollback witness for \(direction.rawValue): the store is at "
                + "offset \(store.nextOffset) / sequence \(store.nextSequence) / "
                + "\(store.attemptsReserved) reserved attempts, but \(witnessLogPath(pairId)) records "
                + "\(w.encryptionNextOffset) / \(w.authenticationNextSequence) / \(w.attemptsReserved). "
                + "That is the signature of a restored or copied store, and going on would reuse "
                + "material this device has already spent. Nothing was burned.")
        }
        // AHEAD is a normal, accepted state — it is what a store looks like after
        // a torn advance or a crash between commit and advance. Only
        // strictly-below refuses.
        return .ok(Self.alignedWith(store, w) ? .aligned : .ahead)
    }

    public func advance(pairId: String, direction: PadDirection,
                        counters: WitnessCounters) throws {
        // Throws on I/O failure. The caller has ALREADY committed, so the output
        // is withheld: the LOSS row. A torn advance under-reports its own value,
        // which is safe only because the operation that tore also errored and
        // withheld its output — so the witness never under-reports below a state
        // whose output was RELEASED, and the next clean advance re-records the
        // high-water.
        try fs.appendFile(witnessLogPath(pairId), encodeWitnessRecord(direction, counters))
    }

    public func report(pairId: String, direction: PadDirection,
                       store: StoreHighWaters) -> WitnessState {
        guard let effective = readEffective(pairId), let w = effective[direction] else {
            return .inconsistent
        }
        if Self.belowWitness(store, w) { return .regressed }
        return Self.alignedWith(store, w) ? .aligned : .ahead
    }
}

/// `fs` is the WITNESS Fs, never the store Fs unless the caller deliberately
/// passes the weak same-domain configuration.
public func witnessFor(fs: Fs, kind: WitnessKind) -> Witness {
    kind == .local ? LocalWitness(fs: fs) : NoneWitness()
}
